libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp

```diff
@@ -227,8 +227,10 @@ bool NavEKF3_core::resetHeightDatum(void)
         }
     }
 
-    // adjust the terrain state
-    terrainState += oldHgt;
+    // set the terrain state to zero (on ground). The adjustment for
+    // frame height will get added in the later constraints
+    terrainState = 0;
```

Comment:
To guard against potential use of this function in a as yet unknow airborne application we should either set the terrain to the default on ground value or if airborne, increment by the amount of the reset, eg:

if (onGround) {
    terrainState = stateStruct.position.z + rngOnGnd;
} else {
    terrainState += oldHgt;
}

Alternatively we could check if on ground at the top of the function, eg:

    if (activeHgtSource == HGT_SOURCE_RNG || !onGround) {
        return false;
    }

Reply:
i've gone for the 2nd option, refusing reset when not onGround